### PR TITLE
Refactor ActionListener#map towards Stricter API

### PR DESCRIPTION
**Pull request by original-brownbear**
Making `#map` look and feel a little nicer, optimize chains of `#map`,
and replace `#delegateFailure` calls with `#map` calls where possible
in order to enforce callbacks not throwing where possible.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/ActionListener.java [hunk 1]
```diff
@@ -45,6 +45,71 @@
...
+     *
+     * @param fn Function to apply to listener response
+     * @param <T> Response type of the wrapped listener
+     * @return a listener that maps the received response and then passes it to this instance
+     */
+    default <T> ActionListener<T> map(CheckedFunction<T, Response, Exception> fn) {
+        return new MappedActionListener<>(fn, this);
+    }
+
+    final class MappedActionListener<Response, O> implements ActionListener<Response> {
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/action/ActionListener.java:66]: Can we use a more descriptive type name than "O"?

**Review by rjernst — APPROVED** [review]
LGTM, one nit

File: server/src/main/java/org/elasticsearch/action/ActionListener.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/action/ActionListener.java]: ++

**Comment by original-brownbear** [comment]
Thanks Ryan!

**Pull request merged**